server/src/test/java/org/elasticsearch/cluster/action/shard/ShardFailedClusterStateTaskExecutorTests.java

```diff
@@ -298,61 +297,29 @@ private ShardRouting nonExistentShardRouting(Index index, List<String> nodeIds,
         );
     }
 
-    private static void assertTasksSuccessful(
-        List<FailedShardUpdateTask> tasks,
-        ClusterStateTaskExecutor.ClusterTasksResult<FailedShardUpdateTask> result,
-        ClusterState clusterState,
-        boolean clusterStateChanged
-    ) {
-        List<Tuple<FailedShardUpdateTask, ClusterStateTaskExecutor.TaskResult>> taskResultList = tasks.stream()
-            .map(t -> Tuple.tuple(t, ClusterStateTaskExecutor.TaskResult.success(NO_OP_TASK_LISTENER)))
-            .collect(Collectors.toList());
-        assertTaskResults(taskResultList, result, clusterState, clusterStateChanged);
-    }
+    private ClusterState executeAndAssertSuccessful(ClusterState clusterState, List<FailedShardUpdateTask> tasks, boolean expectChange)
+        throws Exception {
 
-    private static void assertTaskResults(
-        List<Tuple<FailedShardUpdateTask, ClusterStateTaskExecutor.TaskResult>> taskResultList,
-        ClusterStateTaskExecutor.ClusterTasksResult<FailedShardUpdateTask> result,
-        ClusterState clusterState,
-        boolean clusterStateChanged
-    ) {
-        // there should be as many task results as tasks
-        assertEquals(taskResultList.size(), result.executionResults().size());
-
-        for (Tuple<FailedShardUpdateTask, ClusterStateTaskExecutor.TaskResult> entry : taskResultList) {
-            // every task should have a corresponding task result
-            assertTrue(result.executionResults().containsKey(entry.v1()));
-
-            // the task results are as expected
-            assertEquals(entry.v1().toString(), entry.v2().isSuccess(), result.executionResults().get(entry.v1()).isSuccess());
+        final var resultingState = ClusterStateTaskExecutorUtils.executeAndAssertSuccessful(clusterState, executor, tasks);
+        if (expectChange) {
+            assertNotSame(clusterState, resultingState);
+        } else {
+            assertSame(clusterState, resultingState);
         }
 
-        List<ShardRouting> shards = clusterState.getRoutingTable().allShards();
-        for (Tuple<FailedShardUpdateTask, ClusterStateTaskExecutor.TaskResult> entry : taskResultList) {
-            if (entry.v2().isSuccess()) {
-                // the shard was successfully failed and so should not be in the routing table
-                for (ShardRouting shard : shards) {
-                    if (shard.assignedToNode()) {
-                        assertFalse(
```

Comment:
Is it possible to add some count/size based assertion?

Reply:
Hmm. I tried a bunch of different ideas but found it very tricky to say how many shards we expect here - we start from some random-ish routing table containing unassigned/initializing shards, and then we fail some shards and assign some others.

I pushed 03dba4c0d939cc084a5707d6ed1ae0aa69b52f33 since I noticed that we wouldn't have any started replicas in the tests today.